### ISSUE TITLE
[SecurityHub] Disparity between ResourceCount metric and number of instances in Security Hub

### ISSUE THREAD

**Issue by lanejlanej** [cloud-custodian/cloud-custodian#3759]
I have a simple policy which uses the 'unused' filter on 'security-group' resource with the 'post-finding' action to send to SecurityHub.

However, I have observed, that when the ResourceCount CloudWatch metric for this policy has a value of 10 only one matching security group instance is displayed in Security Hub findings.

**Comment by kapilt** [comment]
are you using a recent custodian, we had to change the batch_size we report findings into security with, as there is a ux bug that only shows a single resource in a finding, even though findings support up to 10 resources in a single finding. you may have to reset/remove the tag on the resource to get a fresh finding. latest custodian does a single finding per resource.

**Comment by lanejlanej** [comment]
Hi, I was using an old version of custodian.  So, I upped the version to 0.8.42.1, removed the FindingId tag from the security groups, archived all mentions of the policy and removed the Cloud Custodian policy (and lambda) via mugc. I then re-added the policy.
The result was the same sort of thing: 11 groups were tagged as reflected in the ResourceCount metric, but only 2 instances were displayed in SecurityHub (The first 2 in the list displayed in AWS console if the FindingId tag column is added to the display, and used as the sort key)

**Comment by lanejlanej** [comment]
This is what gets logged when the lambda runs:
--------------------
[DEBUG]	2019-04-04T09:53:03.396Z	3c6aab98-7dda-4db0-81a2-66695de4138a	Filtered from 19 to 19 securitygroup

[DEBUG]	2019-04-04T09:53:03.396Z	3c6aab98-7dda-4db0-81a2-66695de4138a	sg-perm-refs using 5 sgs, new refs 4 total 8

[DEBUG]	2019-04-04T09:53:03.668Z	3c6aab98-7dda-4db0-81a2-66695de4138a	Filtered from 17 to 17 awslambda

[DEBUG]	2019-04-04T09:53:03.668Z	3c6aab98-7dda-4db0-81a2-66695de4138a	lambdas using 0 sgs, new refs 0 total 8

[DEBUG]	2019-04-04T09:53:03.784Z	3c6aab98-7dda-4db0-81a2-66695de4138a	Filtered from 0 to 0 launchconfig

[DEBUG]	2019-04-04T09:53:03.784Z	3c6aab98-7dda-4db0-81a2-66695de4138a	launch-configs using 0 sgs, new refs 0 total 8

[DEBUG]	2019-04-04T09:53:03.879Z	3c6aab98-7dda-4db0-81a2-66695de4138a	0 of 11 groups w/ peered refs

[DEBUG]	2019-04-04T09:53:03.881Z	3c6aab98-7dda-4db0-81a2-66695de4138a	Filtered from 19 to 11 securitygroup

[INFO]	2019-04-04T09:53:03.881Z	3c6aab98-7dda-4db0-81a2-66695de4138a	policy: security-groups-not-referenced resource:security-group region:eu-west-1 count:11 time:1.20

**Number of resources: 11
[DEBUG]	2019-04-04T09:53:04.341Z	3c6aab98-7dda-4db0-81a2-66695de4138a	policy:security-groups-not-referenced securityhub 2 findings resources 0 new 11 updated 0 failed**

[INFO]	2019-04-04T09:53:04.342Z	3c6aab98-7dda-4db0-81a2-66695de4138a	policy: security-groups-not-referenced action: otherresourcepostfinding resources: 11 execution_time: 0.46

END RequestId: 3c6aab98-7dda-4db0-81a2-66695de4138a
REPORT RequestId: 3c6aab98-7dda-4db0-81a2-66695de4138a	Duration: 2303.47 ms	Billed Duration: 2400 ms 	Memory Size: 512 MB	Max Memory Used: 97 MB

**Comment by lanejlanej** [comment]
Problem seems to be that **multiple distinct security group instances are being tagged with the same FindingId tag**.  So they get grouped together as 2 findings to report.
With a little extra debug:
[DEBUG]	2019-04-04T10:39:47.276Z	f5104089-b1f4-4800-8961-b45bda51015f	finding_tag: eu-west-1/627886280200/0df7bbb7abe181dc7a07f8b0ae7f8069/b3225a896c2cfdf89c370da308ebb95d:2019-04-03T11:45:33.548473+00:00

[DEBUG]	2019-04-04T10:39:47.276Z	f5104089-b1f4-4800-8961-b45bda51015f	GroupId: sg-065e7c42932a3d2b5

[DEBUG]	2019-04-04T10:39:47.276Z	f5104089-b1f4-4800-8961-b45bda51015f	finding_tag: eu-west-1/627886280200/0df7bbb7abe181dc7a07f8b0ae7f8069/b3225a896c2cfdf89c370da308ebb95d:2019-04-03T11:45:33.548473+00:00

[DEBUG]	2019-04-04T10:39:47.276Z	f5104089-b1f4-4800-8961-b45bda51015f	GroupId: sg-072f57e9a3ed7cc86

[DEBUG]	2019-04-04T10:39:47.276Z	f5104089-b1f4-4800-8961-b45bda51015f	finding_tag: eu-west-1/627886280200/0df7bbb7abe181dc7a07f8b0ae7f8069/b3225a896c2cfdf89c370da308ebb95d:2019-04-03T11:45:33.548473+00:00

**Comment by kapilt** [comment]
There is a `batch_size` setting that as of latest version of custodian defaults to 1. For older versions this was 10, that controls the number of resources that we send in a single finding and tag with that finding id. Is it possible after you cleaned up it ended up redeploying the older version? Note the tag on the extant resources would need to be reset. In general when debugging its best to simplify, ie not do this in lambda.

**Comment by lanejlanej** [comment]
You're right batch_size does seem to be the issue. It is getting set to 10 in 0.8.42.1 ->
securityhub.py
191
192         now = datetime.utcnow().replace(tzinfo=tzutc()).isoformat()
193         batch_size = self.data.get('batch_size', 10)
194         stats = Counter()

**Comment by kapilt** [comment]
apologies, I thought the change here had been released, I'll queue up a new one in the next few days.  if you set that explicitly to batch_size: 1 that should resolve.

release tag link to batch_size default
https://github.com/cloud-custodian/cloud-custodian/blob/0.8.42.1/c7n/actions/securityhub.py#L193

**Comment by kapilt** [comment]
added a pr #3765 to cut a release (0.8.43.0)
